### PR TITLE
expression: fix the behavior when adding date with big interval | tidb-test=pr/2260

### LINKED ISSUE #49227: Add date with a big intervals result incorrect values
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

<!-- a step by step guide for reproducing the bug. -->

```
> select "1000-01-01 00:00:00" + INTERVAL 9223372036854775808 day;
> select "1000-01-01 00:00:00" + INTERVAL -9223372036854775809 day;
> create table t(a datetime);
> set sql_mode='';
> insert into t values("1000-01-01 00:00:00" + INTERVAL 18446744073709551616 day);
```

### 2. What did you expect to see? (Required)

In MySQL:

```
mysql> select "1000-01-01 00:00:00" + INTERVAL 9223372036854775808 day;
+----------------------------------------------------------+
| "1000-01-01 00:00:00" + INTERVAL 9223372036854775808 day |
+----------------------------------------------------------+
| NULL                                                     |
+----------------------------------------------------------+
1 row in set, 1 warning (0.00 sec)

mysql> select "1000-01-01 00:00:00" + INTERVAL -9223372036854775809 day;
+-----------------------------------------------------------+
| "1000-01-01 00:00:00" + INTERVAL -9223372036854775809 day |
+-----------------------------------------------------------+
| NULL                                                      |
+-----------------------------------------------------------+
1 row in set, 2 warnings (0.00 sec)

mysql> insert into t values("1000-01-01 00:00:00" + INTERVAL 18446744073709551616 day); -- insert when sql_mode=''
Query OK, 1 row affected, 2 warnings (0.00 sec)

mysql> select * from t;
+------+
| a    |
+------+
| NULL |
+------+
1 row in set (0.01 sec)
```

### 3. What did you see instead (Required)

```
mysql> select "1000-01-01 00:00:00" + INTERVAL 9223372036854775808 day;
+----------------------------------------------------------+
| "1000-01-01 00:00:00" + INTERVAL 9223372036854775808 day |
+----------------------------------------------------------+
| 1000-01-01 00:00:00                                      |
+----------------------------------------------------------+
1 row in set (0.00 sec)

mysql> select "1000-01-01 00:00:00" + INTERVAL -9223372036854775809 day;
+-----------------------------------------------------------+
| "1000-01-01 00:00:00" + INTERVAL -9223372036854775809 day |
+-----------------------------------------------------------+
| 1000-01-01 00:00:00                                       |
+-----------------------------------------------------------+
1 row in set, 1 warning (0.00 sec)

mysql> insert into t values("1000-01-01 00:00:00" + INTERVAL 18446744073709551616 day);
ERROR 1690 (22003): %s value is out of range in '%s'
```

### 4. What is your TiDB version? (Required)

<!-- Paste the output of SELECT tidb_version() -->

master

```
mysql> select tidb_version();
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| tidb_version()                                                                                                                                                                                                                                                             |
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Release Version: v7.6.0-alpha
Edition: Community
Git Commit Hash: 795335a7f88da6197221da145c854c2dbfe22f95
Git Branch: heads/refs/tags/v7.6.0-alpha
UTC Build Time: 2023-12-03 14:26:07
GoVersion: go1.21.4
Race Enabled: false
Check Table Before Drop: false
Store: tikv |
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```

### PR DESCRIPTION
**Pull request by lcwangchao**
### What problem does this PR solve?

Issue Number: close #49227

### What changed and how does it work?

- `date + interval X uint` when X is a big value, we should get `NULL`
- Fix some other interval cases to keep the same behavior with MySQL 8.0:
  - `date + interval "true" DAY` should not add one day, only `date + interval true DAY` should be seen as add 1 time unit.
  - `date + interval "xxx1" uint`, if an interval string does not start with +/- or a digit, it should be truncated as 0, not 1.
  - The unit `SECOND` is special, it should accept string in in exponential format, for example:  `select "1900-01-01 00:00:00" + INTERVAL "1e2" second;` should add 100 seconds.
  - If the unit is `SECOND` , the interval string should be regarded as float, other units should be regarded as int.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
To make some behaviors same with Mysql8.0:

fix the behavior when adding date with big interval
interval with an invalid prefix will be seen as a zero value after this PR
interval with a string "true" will be seen as a zero value after this PR
```

### CODE REVIEW

File: pkg/expression/builtin_time.go [hunk 1]
```diff
@@ -2754,7 +2754,7 @@ type baseDateArithmetical struct {
 
 func newDateArithmeticalUtil() baseDateArithmetical {
 	return baseDateArithmetical{
-		intervalRegexp: regexp.MustCompile(`-?[\d]+`),
+		intervalRegexp: regexp.MustCompile(`^[+-]?[\d]+`),
```
Comment (lcwangchao) [inline pkg/expression/builtin_time.go:2757]: `XXX1` should be parsed as "", not "1"

File: pkg/expression/builtin_time.go [hunk 2]
```diff
@@ -2936,6 +2952,13 @@ func (du *baseDateArithmetical) getIntervalFromInt(ctx sessionctx.Context, args
 	if isNull || err != nil {
 		return "", true, err
 	}
+
+	if mysql.HasUnsignedFlag(args[1].GetType().GetFlag()) && interval < 0 {
+		tc := ctx.GetSessionVars().StmtCtx.TypeCtx()
+		err = tc.HandleTruncate(types.ErrDatetimeFunctionOverflow.GenWithStackByArgs("datetime"))
```
[the remark this answers is not in the document]
Comment (YangKeao) [reply pkg/expression/builtin_time.go]: Or use `ctx.GetSessionVars().StmtCtx.HandleError`? They have nearly the same behavior now :thinking: .

File: pkg/expression/builtin_time.go [hunk 3]
```diff
@@ -2864,17 +2864,55 @@ func (du *baseDateArithmetical) getIntervalFromString(ctx sessionctx.Context, ar
...
-	if toLower := strings.ToLower(unit); toLower == "day" || toLower == "hour" {
-		if strings.ToLower(interval) == "true" {
-			interval = "1"
-		} else if strings.ToLower(interval) == "false" {
+
+	interval, err = du.intervalReformatString(ctx.GetSessionVars().StmtCtx.TypeCtx(), interval, unit)
+	return interval, false, err
+}
+
+func (du *baseDateArithmetical) intervalReformatString(tc types.Context, str string, unit string) (interval string, err error) {
```
Comment (YangKeao) [inline pkg/expression/builtin_time.go:2872]: How about using `errctx`, rather than the `HandleTruncate` in `types.Context`?

**Review by YangKeao — COMMENTED** [review]
Rest LGTM

File: pkg/expression/builtin_time.go (hunk 3 above)
Comment (lcwangchao) [reply pkg/expression/builtin_time.go]: done

**Review by YangKeao — APPROVED** [review]
LGTM

File: pkg/expression/builtin_time.go [hunk 4]
```diff
@@ -2864,17 +2865,58 @@ func (du *baseDateArithmetical) getIntervalFromString(ctx sessionctx.Context, ar
...
+
+	interval, err = du.intervalReformatString(ctx.GetSessionVars().StmtCtx.ErrCtx(), interval, unit)
+	return interval, false, err
+}
+
+func (du *baseDateArithmetical) intervalReformatString(ec errctx.Context, str string, unit string) (interval string, err error) {
+	switch strings.ToUpper(unit) {
+	case "MICROSECOND", "MINUTE", "HOUR", "DAY", "WEEK", "MONTH", "QUARTER", "YEAR":
+		str = strings.TrimSpace(str)
+		// a single unit value has to be specially handled.
```
Comment (windtalker) [inline pkg/expression/builtin_time.go:2877]: The case of `interval true day` will be convert to `interval 1 day` before this function?

File: pkg/expression/builtin_time.go [hunk 5]
```diff
@@ -2936,6 +2977,11 @@ func (du *baseDateArithmetical) getIntervalFromInt(ctx sessionctx.Context, args
 	if isNull || err != nil {
 		return "", true, err
 	}
+
+	if mysql.HasUnsignedFlag(args[1].GetType().GetFlag()) {
```
Comment (windtalker) [inline pkg/expression/builtin_time.go:2981]: When will the type has unsigned flag?

File: pkg/expression/builtin_time.go (hunk 4 above)
Comment (lcwangchao) [reply pkg/expression/builtin_time.go:2877]: The behaviors of handling `interval true day` are the same before/after this PR, the only difference is the behavior to handle the string mode `interval "true"` way.

Before this PR:

```
TiDB root@127.0.0.1:test> select "1900-01-01" + interval "true" day;
+------------------------------------+
| "1900-01-01" + interval "true" day |
+------------------------------------+
| 1900-01-02                         |
+------------------------------------+
```

After this PR:

```
TiDB root@127.0.0.1:test> select "1900-01-01" + interval "true" day;
+------------------------------------+
| "1900-01-01" + interval "true" day |
+------------------------------------+
| 1900-01-01                         |
+------------------------------------+
```

This keeps the same with Mysql8.0

File: pkg/expression/builtin_time.go (hunk 5 above)
Comment (lcwangchao) [reply pkg/expression/builtin_time.go:2981]: For example:

```
select "1900-01-01" + interval cast("18446744073709551615" as unsigned) day;
```

Comment (windtalker) [reply pkg/expression/builtin_time.go:2981]: OK

**Review by windtalker — APPROVED** [review]
LGTM

**Pull request merged**